File: 2024/03/main.py
#!/usr/bin/env python3

import re
import string
import sys

def checkLine(line:str):
    runningSum = 0
    k = 0
    lineLength = len(line)
    while True:
        if k >= lineLength:
            break
        curChar = line[k]
        k+=1
        if curChar != "m":
            continue
        if line[k] != "u":
            continue
        k+=1
        if line[k] != "l":
            continue
        k+=1
        if line[k] != "(":
            continue
        k+=1

        # This is the first number up to 3 digits
        m = 0
        firstMul = ""
        while True:
            if line[k+m] in string.digits:
                firstMul += line[k+m]
                m+=1
            else:
                break
        if line[k+m] != ",":
            continue
        k+=m+1

        # Next number
        m = 0
        secondMul = ""
        while True:
            if line[k+m] in string.digits:
                secondMul += line[k+m]
                m+=1
            else:
                break

        if line[k+m] != ")":
            continue
        k+=m+1

        # We now have firstMul and secondMul values.
        # Now we need to make sure they are up to 3 digits
        if len(firstMul) > 3 or len(secondMul) > 3:
            continue

        runningSum += int(firstMul)*int(secondMul)

    return runningSum

def doPart1():
    filename = "example.dat"
    if len(sys.argv) > 1:
        filename = sys.argv[1]

    with open(filename, "r") as f:
        runningSum = 0
        for li in f:
            line = li.strip()
            if not line:
                continue
            curSum = checkLine(line)
            runningSum += curSum
    print(runningSum)

def checkLineDoDont(line, doMul=True):
    # Guess I'll use regex...
    runningSum = 0
    captureGroups = re.findall("do\\(\\)|don't\\(\\)|mul\\(\\d+,\\d+\\)", line)

    for captureGroup in captureGroups:
        if captureGroup == "do()":
            doMul = True
        elif captureGroup == "don't()":
            doMul = False
        else:
            if doMul:
                a,b = re.findall("\\d+", captureGroup)
                x,y = int(a), int(b)
                runningSum += x*y

    return runningSum, doMul

def doPart2():
    filename = "input.dat"
    if len(sys.argv) > 1:
        filename = sys.argv[1]

    doMul = True
    with open(filename, "r") as f:
        runningSum = 0
        for li in f:
            line = li.strip()
            if not line:
                continue
            curSum, doMul = checkLineDoDont(line, doMul)
            runningSum += curSum
    print(runningSum)

def main():
    doPart1()
    doPart2()

if __name__ == "__main__":
    main()

File: 2024/03/main.cpp

#include "../helpers/extras.hh"
#include <array>
#include <regex>

void doPart1(const char* filename)
{
    std::ifstream file(filename);
    std::string line;
    int sum = 0;
    while (getline(file, line)) {
        stripString(line);
        std::smatch match;
        std::smatch vals;
        std::regex r("mul\\(\\d+,\\d+\\)");
        std::regex nums("\\d+");
        while (regex_search(line, match, r)) {
            std::string valMatch = match.str(0);
            std::array<int,2> ab;
            int k = 0;
            while (regex_search(valMatch, vals, nums)) {
                std::string a = vals.str(0);
                ab[k++] = std::atoi(a.c_str());
                valMatch = vals.suffix().str();
            }
            sum += ab[0]*ab[1];
            line = match.suffix().str();
        }

    }
    file.close();
    printf("%d\n", sum);
}

void doPart2(const char* filename)
{
    std::ifstream file(filename);
    std::string line;
    int sum = 0;
    bool doMul = true;
    while (getline(file, line)) {
        stripString(line);
        std::smatch match;
        std::smatch vals;
        std::regex r("mul\\(\\d+,\\d+\\)|do\\(\\)|don't\\(\\)");
        std::regex nums("\\d+");
        while (regex_search(line, match, r)) {
            std::string valMatch = match.str(0);
            if (valMatch == "do()") doMul = true;
            else if (valMatch == "don't()") doMul = false;
            else {
                if (doMul) {
                    std::array<int,2> ab;
                    int k = 0;
                    while (regex_search(valMatch, vals, nums)) {
                        std::string a = vals.str(0);
                        ab[k++] = std::atoi(a.c_str());
                        valMatch = vals.suffix().str();
                    }
                    sum += ab[0]*ab[1];
                }
            }
            line = match.suffix().str();
        }

    }
    file.close();
    printf("%d\n", sum);
}

int main(int narg, char* args[])
{
    std::string filename = "example.dat";
    if (narg > 1) filename = std::string(args[1]);
    doPart1(filename.c_str());
    doPart2(filename.c_str());
    return 0;
}
